include/oneapi/dpl/internal/random_impl/extreme_value_distribution.h

```diff
@@ -0,0 +1,198 @@
+        scalar_type a;
+        scalar_type b;
+    };
+
+    // Constructors
+    extreme_value_distribution() : extreme_value_distribution(scalar_type{0.0}) {}
+    explicit extreme_value_distribution(scalar_type __a, scalar_type __b = scalar_type{1.0}) : a_(__a), b_(__b) {}
+    explicit extreme_value_distribution(const param_type& __params) : a_(__params.a), b_(__params.b) {}
+
+    // Reset function
+    void
+    reset()
+    {
+    }
+
+    // Property functions
+    scalar_type
+    a() const
+    {
+        return a_;
+    }
+
+    scalar_type
+    b() const
+    {
+        return b_;
+    }
+
+    param_type
+    param() const
+    {
+        return param_type(a_, b_);
+    }
+
+    void
+    param(const param_type& __param)
+    {
+        a_ = __param.a;
+        b_ = __param.b;
+    }
+
+    scalar_type
+    min() const
+    {
+        return -std::numeric_limits<scalar_type>::max();
+    }
+
+    scalar_type
+    max() const
+    {
+        return std::numeric_limits<scalar_type>::max();
+    }
+
+    // Generate functions
+    template <class _Engine>
+    result_type
+    operator()(_Engine& __engine)
+    {
+        return operator()<_Engine>(__engine, param_type(a_, b_));
+    }
+
+    template <class _Engine>
+    result_type
+    operator()(_Engine& __engine, const param_type& __params)
+    {
+        return generate<size_of_type_, _Engine>(__engine, __params);
+    }
+
+    template <class _Engine>
+    result_type
+    operator()(_Engine& __engine, unsigned int __random_nums)
+    {
+        return operator()<_Engine>(__engine, param_type(a_, b_), __random_nums);
+    }
+
+    template <class _Engine>
+    result_type
+    operator()(_Engine& __engine, const param_type& __params, unsigned int __random_nums)
+    {
+        return result_portion_internal<size_of_type_, _Engine>(__engine, __params, __random_nums);
+    }
+
+  private:
+    // Size of type
+    static constexpr int size_of_type_ = internal::type_traits_t<result_type>::num_elems;
+
+    // Static asserts
+    static_assert(::std::is_floating_point<scalar_type>::value,
+                  "oneapi::dpl::extreme_value_distribution. Error: unsupported data type");
+
+    // Distribution parameters
+    scalar_type a_;
+    scalar_type b_;
+
+    // Implementation for generate function
+    template <int _Ndistr, class _Engine>
+    typename ::std::enable_if<(_Ndistr != 0), result_type>::type
+    generate(_Engine& __engine, const param_type& __params)
+    {
+        return generate_vec<_Ndistr, _Engine>(__engine, __params);
+    }
+
+    // Specialization of the scalar generation
+    template <int _Ndistr, class _Engine>
+    typename ::std::enable_if<(_Ndistr == 0), result_type>::type
+    generate(_Engine& __engine, const param_type& __params)
+    {
+        oneapi::dpl::exponential_distribution<scalar_type> __e;
+        return __params.a - __params.b * sycl::log(__e(__engine));

```

Comment:
could we have log(0) here?

Reply:
It possible as I see. I added the callback.